test/test_nn.py

```diff
@@ -11444,10 +11517,88 @@ def __init__(self):
+            oc_g = combination[3]
+            p = combination[4]
+            d = combination[5]
+            s = combination[6]
+            k = combination[7]
+            osize = combination[8]
+            has_bias = combination[9]
+
+            isize = []
+            for i in range(len(osize)):
+                isize.append(s[i] * (osize[i] - 1) + 1 + d[i] * (k[i] - 1) - 2 * p[i])
+
+            if isize[0] > 0 and isize[1] > 0:
+                ic, oc = ic_g * group, oc_g * group
+                torch.manual_seed(1)
+                input = torch.randn((bs, ic, isize[0], isize[1]), device=device, dtype=dtype, requires_grad=True)
+                torch.manual_seed(2)
+                weight = torch.randn((oc, ic_g, k[0], k[1]), device=device, dtype=dtype) * 0.01
+                torch.manual_seed(3)
+                bias = None if has_bias else torch.randn((oc), device=device, dtype=dtype)
+
+                thnn_output = self._thnn_conv_group(input, weight, k, bias, s, p, d, group)
+                output = torch.conv2d(input, weight, bias, s, p, d, group)
+                if self.device_type == 'cuda':
+                    self.assertAlmostEqual(
+                        output, thnn_output,
+                        msg='device:{}, dtype:{}, input size:{}, group:{}, '
+                            'batchsize:{}, input channel:{}, output channel:{}, '
+                            'bias:{}, padding:{}, dilation:{}, stride:{}, '
+                            'kernel:{}'.format(device, dtype, isize, group, bs, ic,
+                                               oc, has_bias, p, d, s, k),
+                        delta=1e-2
+                    )
+                else:
+                    self.assertEqual(
+                        output, thnn_output,
+                        msg='device:{}, dtype:{}, input size:{}, group:{}, '
+                            'batchsize:{}, input channel:{}, output channel:{}, '
+                            'bias:{}, padding:{}, dilation:{}, stride:{}, '
+                            'kernel:{}'.format(device, dtype, isize, group, bs, ic,
+                                               oc, has_bias, p, d, s, k)
+                    )
 
+                if self.device_type == 'cpu' and torch.backends.mkldnn.is_available():
+                    mkldnn_output = torch.mkldnn_convolution(input, weight, bias, p, s, d, group)
+                    self.assertEqual(
+                        mkldnn_output, thnn_output,
+                        msg='device:{}, dtype:{}, input size:{}, group:{}, '
+                            'batchsize:{}, input channel:{}, output channel:{}, '
+                            'bias:{}, padding:{}, dilation:{}, stride:{}, '
+                            'kernel:{}'.format(device, dtype, isize, group, bs, ic,
+                                               oc, has_bias, p, d, s, k)
+                    )
+
+                if self.device_type == 'cuda' and torch.backends.cudnn.is_available():
```

Comment:
If cudnn backend is available, you've already tested this when you were comparing conv2d with thnn output. If it's not available  then that previous test serves no purpose. Same goes for mkl-specific testing.

Reply:
This test does not have a good structure - it mixes comparing thnn agains mkldnn and cudnn in a single test, making this test hard to follow and doing extra work, as pointed out above. 
You may think that you are saving computation of thnn_output for both mkldnn and cudnn, but that's not the case - 2 different tests will be generated for cuda and cpu.